mesh/meshdb.go

```diff
@@ -137,7 +148,46 @@ func (m *DB) Transactions() database.Getter {
 
 // AddBallot adds a ballot to the database.
 func (m *DB) AddBallot(b *types.Ballot) error {
-	return m.addBallot(b)
+	mal, err := identities.IsMalicious(m.db, b.SmesherID().Bytes())
+	if err != nil {
+		return err
+	}
+	if mal {
+		b.SetMalicious()
+	}
+
+	// it is important to run add ballot and set identity to malicious atomically
+	tx, err := m.db.Tx(context.Background())
+	if err != nil {
+		return err
+	}
+	defer tx.Release()
+
+	if err := ballots.Add(tx, b); err != nil && !errors.Is(err, sql.ErrObjectExists) {
+		return err
+	}
+	if !mal {
+		count, err := ballots.CountByPubkeyLayer(tx, b.LayerIndex, b.SmesherID().Bytes())
+		if err != nil {
+			return err
+		}
+		if count > 1 {
+			if err := identities.SetMalicious(tx, b.SmesherID().Bytes()); err != nil {
```

Comment:
it feels slightly out of place to set an identity malicious at the DB layer.
thinking forward, how should this information gets propagated up to proceed with blacklisting.

maybe it make more sense to add a malicious field in the ballot table? we can mark that this ballot is determined to use eligibility more than once and propagate that error up. let the upstream caller to decide whether to mark an identity malicious and further blacklist? (and perhaps retroactively mark other ballots of the same identity malicious)

Reply:
> thinking forward, how should this information gets propagated up to proceed with blacklisting.

it is blacklisted by tracking that identity is malicious. what exactly is meant by blacklisting?

> maybe it make more sense to add a malicious field in the ballot table?

it is hard for me to tell if it works or not. there is an important concurrency issue that i addressed by executing statements in transaction. we need to guarantee that if there are 2 ballots in the datatabse then identity is set to malicious. 

for example if Count statement is executed not within transaction then when we are adding two distinct ballots they both may return 0 and identity won't be malicious.

other case, if SetMalicious is not within transaction then ballots.Add may succeed and SetMalicious fail - identity again not tracked as malicious.

Reply:
> maybe it make more sense to add a malicious field in the ballot table?

recalled another reason. if identity was malicious at some layer in the past we don't want to use new ballots from it. scanning whole ballots table to decide if identity is malicious is not an option

Reply:
> it feels slightly out of place to set an identity malicious at the DB layer.

also I think that we don't need MeshDB at all now, low level db stuff is moved to sql module.
and higher level composition should be moved to the Mesh

Reply:
> it is blacklisted by tracking that identity is malicious. what exactly is meant by blacklisting?

from one of the R&D recording, Tal said we need to gossip maleficence proof on a bad actor. i'd imagine once an identity is blacklisted, we don't consider data from them anymore. but i don't think there is an official definition.

> for example if Count statement is executed not within transaction then when we are adding two distinct ballots they both may return 0 and identity won't be malicious.

right. that's why i said earlier that Count and Add needs to be in a DBTX (db transaction)

> recalled another reason. if identity was malicious at some layer in the past we don't want to use new ballots from it. scanning whole ballots table to decide if identity is malicious is not an option

i am not saying to remove malicious field from identities, i am saying to add another malicious field to ballots. that way we know at the time of add this ballot is determined malicious.

as your implementation now, whether a ballot is malicious actually changes. for example, ballot A is added, then an identity is determined malicious when ballot B is added, both ballot A and ballot B are queried as malicious later. this is confusing for debugging, especially if ballot A is chosen as base ballot, but is actually malicious.

Reply:
> i am not saying to remove malicious field from identities, i am saying to add another malicious field to ballots. that way we know at the time of add this ballot is determined malicious.

ballot can be determined as malicious only after we saw two ballots with the same eligibility. once we know that identity is malicious - all ballots from this identity are malicious. it should be expected that some ballots, due to the concurrency or malicious intent, can be used as base ballot before they are discovered as malicious. we can't remove malicious ballots from state but we also don't count weight from them...